src/parsers.rs

```diff
@@ -1,22 +1,784 @@
+                .with_message("Time zone annotation is required for parsing a zoned date time."));
+        }
+
+        if let Some(date_record) = record.date {
+            validate_date_record_impl(date_record).map_err(|e| e.into_temporal_error())?;
+        }
+        if let Some(time_record) = record.time {
+            validate_time_record_impl(time_record).map_err(|e| e.into_temporal_error())?;
+        }
+
+        Ok(record)
+    }
+
+    fn parse_instant_utf16(&self, source: &'a [u16]) -> TemporalResult<IxdtfParseInstantRecord> {
+        let record = self.parse_ixdtf_utf16(source, ParseVariant::DateTime)?;
+
+        let IxdtfParseRecord {
+            date: Some(date),
+            time: Some(time),
+            offset: Some(offset),
+            ..
+        } = record
+        else {
+            return Err(
+                TemporalError::range().with_message("Required fields missing from Instant string.")
+            );
+        };
+
+        validate_date_record_impl(date).map_err(|e| e.into_temporal_error())?;
+        validate_time_record_impl(time).map_err(|e| e.into_temporal_error())?;
+
+        Ok(IxdtfParseInstantRecord { date, time, offset })
+    }
+
+    fn parse_time_utf16(&self, source: &'a [u16]) -> TemporalResult<TimeRecord> {
+        let time_record = self.parse_ixdtf_utf16(source, ParseVariant::Time);
+
+        let Err(ref e) = time_record else {
+            return time_record.and_then(|record| self.check_time_record_utf16(record));
+        };
+
+        let dt_parse = self.parse_date_time_utf16(source);
+
+        match dt_parse {
+            Ok(dt) => self.check_time_record_utf16(dt),
+            _ => Err(TemporalError::range().with_message(format!("{e}"))),
+        }
+    }
+
+    fn parse_year_month_utf16(
+        &self,
+        source: &'a [u16],
+    ) -> TemporalResult<IxdtfParseRecord<'a, Utf16>> {
+        let ym_record = self.parse_ixdtf_utf16(source, ParseVariant::YearMonth);
+
+        let Err(ref e) = ym_record else {
+            return ym_record.and_then(|record| self.check_offset_utf16(record));
+        };
+
+        let dt_parse = self.parse_date_time_utf16(source);
+
+        match dt_parse {
+            Ok(dt) => self.check_offset_utf16(dt),
+            _ => Err(TemporalError::range().with_message(format!("{e}"))),
+        }
+    }
+
+    fn parse_month_day_utf16(
+        &self,
+        source: &'a [u16],
+    ) -> TemporalResult<IxdtfParseRecord<'a, Utf16>> {
+        let md_record = self.parse_ixdtf_utf16(source, ParseVariant::MonthDay);
+
+        let Err(ref e) = md_record else {
+            return md_record.and_then(|record| self.check_offset_utf16(record));
+        };
+
+        let dt_parse = self.parse_date_time_utf16(source);
+
+        match dt_parse {
+            Ok(dt) => self.check_offset_utf16(dt),
+            _ => Err(TemporalError::range().with_message(format!("{e}"))),
+        }
+    }
+
+    fn parse_ixdtf_utf16(
```

Comment:
observation: I also think this code should be shareable.

I think all of this can be made to operate on an `IxdtfParser`, and the end functions construct an `IxdtfParser::from_whatever` and pass it down.

Hell, I think the IxdtfParser can probably just be a field in `TemporalParser`.